### PR TITLE
Use map middleware for custom user agent

### PR DESCRIPTION
**Pull request by guyilin-amazon**
*Issue #, if available:*
#331 #196 

*Description of changes:*
Use map middleware for custom user agent, ref: https://gist.github.com/rcoh/14668328600aa8989b679e42ee92d45f

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by allimn — APPROVED** [review]

File: src/awssdk_driver.rs [hunk 1]
```diff
@@ -21,10 +19,35 @@ use aws_sdk_qldbsession::{
...
+use http::header::HeaderName;
+use http::Uri;
+use tower::ServiceBuilder;
 
 use crate::{error, settings::Environment};
 
+#[derive(Clone, Debug)]
+struct UserAgent;
+
+impl MapRequest for UserAgent {
```
Comment (marcbowes) [inline src/awssdk_driver.rs:33]: Can you please write a test for this that shows that, at least, the code doesn't panic? (Given it uses unwrap.)

File: src/awssdk_driver.rs [hunk 2]
```diff
@@ -21,10 +19,35 @@ use aws_sdk_qldbsession::{
...
+    type Error = Box<dyn Error + Send + Sync + 'static>;
+
+    fn apply(&self, request: operation::Request) -> Result<operation::Request, Self::Error> {
+        request.augment(|mut req, _conf| {
+            req.headers_mut().append(
+                HeaderName::from_static("x-amz-qldb-driver-version"),
+                amazon_qldb_driver::version().to_string().parse().unwrap(),
```
Comment (marcbowes) [inline src/awssdk_driver.rs:40]: I'm a little unsure what's going on here.

https://docs.rs/http/0.2.6/http/header/struct.HeaderMap.html#method.append says the args are generic over K and V. So, what is V? i.e. what is the `parse()` doing to the string?

It seems to me that we shouldn't be doing allocation (heap allocation for the string) plus parse work on every call. Furthermore, having `unwrap` makes me a little nervous as libraries should never panic as doing so can bring down a customer application.

**Review by marcbowes — CHANGES_REQUESTED** [review]

File: src/awssdk_driver.rs (hunk 2 above)
Comment (marcbowes) [reply src/awssdk_driver.rs:40]: I see, so I think you're trying to use `FromStr` to get a `HeaderValue`. Is that right? If so, did you consider https://docs.rs/http/0.2.6/http/header/struct.HeaderValue.html#method.from_static which is a const fn?

Comment (marcbowes) [reply src/awssdk_driver.rs:40]:
```suggestion
                HeaderValue::from_static(amazon_qldb_driver::version())
```

File: src/awssdk_driver.rs [hunk 3]
```diff
@@ -21,10 +19,35 @@ use aws_sdk_qldbsession::{
...
+            );
+            req.headers_mut().append(
+                HeaderName::from_static("x-amz-qldb-shell-version"),
+                env!("CARGO_PKG_VERSION").parse().unwrap(),
```
Comment (marcbowes) [inline src/awssdk_driver.rs:44]:
```suggestion
                HeaderValue::from_static(env!("CARGO_PKG_VERSION")),
```

File: src/awssdk_driver.rs [hunk 4]
```diff
@@ -21,10 +19,35 @@ use aws_sdk_qldbsession::{
 use aws_smithy_client::bounds::SmithyConnector;
 use aws_smithy_client::erase::DynConnector;
 use aws_smithy_client::Client;
-use aws_types::app_name::AppName;
+use aws_smithy_http::{middleware::MapRequest, operation};
+use aws_smithy_http_tower::map_request::MapRequestLayer;
+use http::header::HeaderName;
+use http::Uri;
```
Comment (marcbowes) [inline src/awssdk_driver.rs:25]:
```suggestion
use http::{HeaderValue, Uri};
```

File: src/awssdk_driver.rs (hunk 1 above)
Comment (marcbowes) [reply src/awssdk_driver.rs:33]: If you accept the `from_static` changes, you can ignore this comment.

Comment (guyilin-amazon) [reply src/awssdk_driver.rs:33]: thanks, changes accepted

**Pull request merged**